File: Task1/race_condition.py
"""
ST5004CEM Operating Systems and Security
Task 1 - Race Condition Demonstration

Author: Trisha Rajbanshi
"""

import threading
import time

counter = 0


def increment():
    global counter

    for _ in range(10000):
        temp = counter
        temp += 1
        counter = temp


def main():

    global counter

    print("=" * 60)
    print("RACE CONDITION DEMONSTRATION")
    print("=" * 60)

    threads = []

    for i in range(5):
        thread = threading.Thread(target=increment)
        threads.append(thread)
        thread.start()

    for thread in threads:
        thread.join()

    print(f"Expected Counter Value : 50000")
    print(f"Actual Counter Value   : {counter}")
    print("=" * 60)


if __name__ == "__main__":
    main()

File: Task1/race_condition.c
#include <stdio.h>
#include <pthread.h>

#define NUM_THREADS 3
#define ITERATIONS 100000

int counter = 0;

void *increment(void *arg)
{
    int id = *(int *)arg;

    for(int i = 0; i < ITERATIONS; i++)
    {
        // No mutex protection
        counter++;
    }

    printf("Thread %d completed.\n", id);

    pthread_exit(NULL);
}

int main()
{
    pthread_t threads[NUM_THREADS];
    int ids[NUM_THREADS];

    printf("===== Race Condition Demonstration =====\n\n");

    for(int i = 0; i < NUM_THREADS; i++)
    {
        ids[i] = i + 1;
        pthread_create(&threads[i], NULL, increment, &ids[i]);
    }

    for(int i = 0; i < NUM_THREADS; i++)
    {
        pthread_join(threads[i], NULL);
    }

    printf("\nExpected Counter Value : %d\n", NUM_THREADS * ITERATIONS);
    printf("Actual Counter Value   : %d\n", counter);

    return 0;
}
